server/sonar-server/src/main/java/org/sonar/server/authentication/SsoAuthenticator.java

```diff
@@ -0,0 +1,182 @@
+
+  private static final String NAME_HEADER_PARAM = "sonar.sso.nameHeader";
+  private static final String NAME_HEADER_DEFAULT_VALUE = "X-Forwarded-Name";
+
+  private static final String EMAIL_HEADER_PARAM = "sonar.sso.emailHeader";
+  private static final String EMAIL_HEADER_DEFAULT_VALUE = "X-Forwarded-Email";
+
+  private static final String GROUPS_HEADER_PARAM = "sonar.sso.groupsHeader";
+  private static final String GROUPS_HEADER_DEFAULT_VALUE = "X-Forwarded-Groups";
+
+  private static final String REFRESH_INTERVAL_PARAM = "sonar.sso.refreshIntervalInMinutes";
+  private static final String REFRESH_INTERVAL_DEFAULT_VALUE = "5";
+
+  private static final String LAST_REFRESH_TIME_TOKEN_PARAM = "ssoLastRefreshTime";
+
+  private static final Map<String, String> DEFAULT_VALUES_BY_PARAMETERS = ImmutableMap.of(
+    LOGIN_HEADER_PARAM, LOGIN_HEADER_DEFAULT_VALUE,
+    NAME_HEADER_PARAM, NAME_HEADER_DEFAULT_VALUE,
+    EMAIL_HEADER_PARAM, EMAIL_HEADER_DEFAULT_VALUE,
+    GROUPS_HEADER_PARAM, GROUPS_HEADER_DEFAULT_VALUE,
+    REFRESH_INTERVAL_PARAM, REFRESH_INTERVAL_DEFAULT_VALUE);
+
+  private final System2 system2;
+  private final Settings settings;
+  private final UserIdentityAuthenticator userIdentityAuthenticator;
+  private final JwtHttpHandler jwtHttpHandler;
+
+  public SsoAuthenticator(System2 system2, Settings settings, UserIdentityAuthenticator userIdentityAuthenticator, JwtHttpHandler jwtHttpHandler) {
+    this.system2 = system2;
+    this.settings = settings;
+    this.userIdentityAuthenticator = userIdentityAuthenticator;
+    this.jwtHttpHandler = jwtHttpHandler;
+  }
+
+  public Optional<UserDto> authenticate(HttpServletRequest request, HttpServletResponse response) {
+    if (!settings.getBoolean(ENABLE_PARAM)) {
+      return Optional.empty();
+    }
+    Map<String, String> headerValuesByNames = getHeaders(request);
+    String login = getHeaderValue(headerValuesByNames, LOGIN_HEADER_PARAM);
+    if (login == null) {
+      return Optional.empty();
+    }
+    Optional<UserDto> user = getUserFromToken(request, response);
+    if (user.isPresent() && login.equals(user.get().getLogin())) {
+      return user;
+    }
+
+    UserDto userDto = doAuthenticate(headerValuesByNames, login);
+    jwtHttpHandler.generateToken(userDto, ImmutableMap.of(LAST_REFRESH_TIME_TOKEN_PARAM, system2.now()), request, response);
+    return Optional.of(userDto);
+  }
+
+  private Optional<UserDto> getUserFromToken(HttpServletRequest request, HttpServletResponse response) {
+    Optional<JwtHttpHandler.Token> token = jwtHttpHandler.getToken(request, response);
+    if (!token.isPresent()) {
+      return Optional.empty();
+    }
+    Date now = new Date(system2.now());
+    int refreshIntervalInMinutes = Integer.parseInt(getSettingValue(REFRESH_INTERVAL_PARAM));
+    Long lastFreshTime = (Long) token.get().getProperties().get(LAST_REFRESH_TIME_TOKEN_PARAM);
+    // TODO test when token exists without sso refresh time
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)